Adding u64 support and modify readstring

abadaa opened the pull request:
Add Writer and Reader for U64 (longs)

Modify readString() to accept Charset if UTF-8 isn't the desired encoding

Write tests

pfrank13 commented on polkaj-scale/src/main/java/io/emeraldpay/polkaj/scale/ScaleCodecWriter.java:
```diff
@@ -96,6 +97,10 @@ public void writeUint32(long value) throws IOException {
         ULONG32.write(this, value);
     }
 
+    public void writeUint64(long value) throws IOException {
```
Should take in a BigInteger because long is a signed value

pfrank13 commented on polkaj-scale/src/main/java/io/emeraldpay/polkaj/scale/reader/UInt64Reader.java:
```diff
@@ -0,0 +1,25 @@
+public class UInt64Reader implements ScaleReader<Long> {
+    public static final int SIZE_BYTES = 8;
+
+    public static void reverse(byte[] value) {
```
UInt64 and 128 can share this method if they pass in their SIZE_BYTES, so this can be refactored into a super class or the instance member SIZE_BYTES can be made a non static value that the constructor sets as a subclass for UInt64 and 128

abadaa replied:
So rather than two separate classes (UInt64Reader and UInt128Reader) just make one class? What would be a good name for that?

pfrank13 commented on polkaj-scale/src/main/java/io/emeraldpay/polkaj/scale/writer/ULong64Writer.java:
```diff
@@ -0,0 +1,31 @@
+
+public class ULong64Writer implements ScaleWriter<Long> {
+    @Override
+    public void write(ScaleCodecWriter wrt, Long value) throws IOException {
```
This needs to take in BigInteger as a Long is signed

pfrank13 commented on polkaj-scale/src/main/java/io/emeraldpay/polkaj/scale/writer/ULong64Writer.java:
```diff
@@ -0,0 +1,31 @@
+        if (value < 0) {
+            throw new IllegalArgumentException("Negative numbers are not supported by Uint128");
+        }
+        byte[] array = ByteBuffer.allocate(Long.SIZE / Byte.SIZE).putLong(value).array();
```
Removing long you want something else here, presumably SIZE_BYTES like on the reader since BigIntegers can be anything

pfrank13 commented on polkaj-scale/src/main/java/io/emeraldpay/polkaj/scale/writer/ULong64Writer.java:
```diff
@@ -0,0 +1,31 @@
+        }
+        byte[] encoded = new byte[8];
+        System.arraycopy(array, pos, encoded, encoded.length - len, len);
+        UInt128Reader.reverse(encoded);
```
These are not something that is equivalent as the SIZE_BYTES of a UInt128 is not the same as this

pfrank13 commented on polkaj-scale/src/test/groovy/io/emeraldpay/polkaj/scale/ScaleCodecReaderSpec.groovy:
```diff
@@ -23,6 +24,15 @@ class ScaleCodecReaderSpec extends Specification {
 
+        def "Reads unsigned 64-bit long"() {
+        when:
+        def codec = new ScaleCodecReader(Hex.decodeHex("2a00000000000000"))
```
Change this test data to be FFFFFFFFFFFFFFFF and the value to be 18446744073709551615 you will find this is -1 because you are using a long which is why we can't use them

pfrank13 commented on polkaj-scale/src/test/groovy/io/emeraldpay/polkaj/scale/ScaleCodecWriterSpec.groovy:
```diff
@@ -68,6 +68,14 @@ class ScaleCodecWriterSpec extends Specification {
 
+    def "Write unsigned 64-bit long"() {
+        when:
+        codec.writeUint64(2147483647L)
```
This is within the range of a Long, using this 18446744073709551615 you'll never get there

pfrank13 commented on polkaj-scale/src/test/groovy/io/emeraldpay/polkaj/scale/reader/UInt64Reader.groovy:
```diff
@@ -0,0 +1,63 @@
+
+    def "Reads large number"() {
+        when:
+        def codec = new ScaleCodecReader(Hex.decodeHex("ffffff7f00000000"))
```
Use FFFFFFFFFFFFFFFF

pfrank13 commented on polkaj-scale/src/test/groovy/io/emeraldpay/polkaj/scale/writer/ULong64WriterSpec.groovy:
```diff
@@ -0,0 +1,90 @@
+        codec.write(writer, x)
+        def act = buf.toByteArray()
+        then:
+        Hex.encodeHexString(act) == "ffffff0000000000"
```
Use FFFFFFFFFFFFFFFF

pfrank13 commented on polkaj-scale/src/test/groovy/io/emeraldpay/polkaj/scale/writer/ULong64WriterSpec.groovy:
```diff
@@ -0,0 +1,90 @@
+        codec.write(writer, x)
+        def act = buf.toByteArray()
+        then:
+        Hex.encodeHexString(act) == "ffffffffffffff7f"
```
Use FFFFFFFFFFFFFFFF

pfrank13 commented on polkaj-scale/src/test/groovy/io/emeraldpay/polkaj/scale/writer/ULong64WriterSpec.groovy:
```diff
@@ -0,0 +1,90 @@
+        where:
+        //MAX Integer is 0x3f_ff_ff_ff
+        encoded     | value
+        "000000000000ff00"  | 0x00_ff_00_00_00_00_00_00
```
Add FFFFFFFFFFFFFFFF somewheer

pfrank13 reviewed (changes requested):
Big takeaway is you need to make things read and write BigIntegers and should refactor UInt64 and 128 Reader to have a superclass that allows the subclass to set he number of bytes for the reverse() method. Something like that. Remove Long from that other filename and "unsigned long" is a UInt64 by definition ;)

abadaa commented:
@pfrank13 Added a superclass called BigIntReader to handle the reverse as well as the SIZE_BYTES integers, changed the longs to BigIntegers, and modified the tests appropriately

Ajaxan commented on polkaj-scale/src/main/java/io/emeraldpay/polkaj/scale/writer/UInt64Writer.java:
```diff
@@ -0,0 +1,31 @@
+            pos++;
+        }
+        int len = array.length - pos;
+        if (len > 8) {
```
Probably replace all 8s with `BigIntReader.UINT64_SIZE_BYTES`

abadaa replied:
good catch, forgot to change those after testing

Ajaxan commented on polkaj-scale/src/main/java/io/emeraldpay/polkaj/scale/reader/UInt128Reader.java:
```diff
@@ -5,23 +5,14 @@
     @Override
     public BigInteger read(ScaleCodecReader rdr) {
-        byte[] value = rdr.readByteArray(SIZE_BYTES);
-        reverse(value);
+        byte[] value = rdr.readByteArray(bir.UINT128_SIZE_BYTES);
+        bir.reverse(value);
         return new BigInteger(1, value);
     }
```
I'm only seeing a small slice of the project, so might be totally wrong on this one. Could you have the base class BigIntReader implement the ScaleReader and move the read method override there to reuse less code?

abadaa replied:
I think I'll leave it as is for now (unless Peter thinks it's a good idea to change it) since when it comes to implementing these classes in our project, it's more intuitive to call the UInt64Reader/Writer and UInt128Reader/Writer separately

Ajaxan commented on polkaj-scale/src/main/java/io/emeraldpay/polkaj/scale/writer/UInt64Writer.java:
```diff
@@ -0,0 +1,31 @@
+
+public class UInt64Writer implements ScaleWriter<BigInteger> {
+    @Override
+    public void write(ScaleCodecWriter wrt, BigInteger value) throws IOException {
```
I mentioned above moving the read method to the shared BigIntReader. If that makes sense perhaps do the same thing for the writers along with the write method?

pfrank13 replied:
The writers should just use constructor injection for the only thing varying which is the number of bytes. Then subclasses can pass in the length that pertains to them.

Ajaxan reviewed (commented):
A few potential suggestions given my limited knowledge of the project. I should mention my eyes glazed over a bit on the groovy tests since I've never done them before. The coverage looked good as far as I could understand though.

pfrank13 commented on polkaj-scale/src/main/java/io/emeraldpay/polkaj/scale/reader/UInt128Reader.java:
```diff
@@ -5,23 +5,14 @@
-            value[i] = tmp;
-        }
-    }
+    BigIntReader bir = new BigIntReader();
```
This is redundant because the reverse method is a static. Just call it directly and get rid of this.

pfrank13 commented on polkaj-scale/src/main/java/io/emeraldpay/polkaj/scale/reader/UInt128Reader.java:
```diff
@@ -5,23 +5,14 @@
+    BigIntReader bir = new BigIntReader();
 
     @Override
     public BigInteger read(ScaleCodecReader rdr) {
```
You should refactor this to use constructor injection of the number of bytes a particular read needs and the subclasses can pass in the number of bytes that pertain to them
